### PR TITLE
feat: initial implementation

### PR DESCRIPTION
**Pull request by ivolimasilva**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "@moxy/next-runtime-env",
+  "version": "0.0.0",
```
Comment (ivolimasilva) [inline package.json:3]: @satazor should we start the version at `1.0.0` and publish that or should we have a commit for the version bump that is used to publish to npm?

Comment (satazor) [reply package.json:3]: Let's start with v0.0.0 and use `npm run release -- --release-as 1.0.0` to bump, which in turn uses `standard-version`

File: README.md [hunk 2]
```diff
@@ -1,2 +1,61 @@
 # next-runtime-env
-NextJS plugin to pass environment variables to NextJS's runtime configuration
+NextJS plugin to pass environment variables to NextJS's runtime configuration.
+
+By default, NextJS does not make available `process.env` to the client side. The available solution for this problem is to configure a runtime config in `next.config.js` to pass values to either server and client or just server.
```
Comment (satazor) [inline README.md]: "is to... to pass.."

Can you rephrase this a little bit to avoid having repetitive words?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,61 @@
...
+
+The way to do it is by passing values to the `publicRuntimeConfig` and `serverRuntimeConfig` properties of the NextJS' config.
+
+This plugin does just that.
```
Comment (satazor) [inline README.md]: > This plugin does that automatically for all environment variables starting with a certain prefix.

Thoughts?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,61 @@
...
+
+## Installation
+
+```sh
+$ npm i -D @moxy/next-runtime-env
```
Comment (satazor) [inline README.md]: Nifty pick: can you change `-D` to `--save-dev`? Most devs don't know that alias.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,61 @@
...
+```
+
+This library is written and published in modern JavaScript. If you target older versions of node please make sure to transpile accordingly.
```
Comment (satazor) [inline README.md]: This is not necessary as this library is not part of the client-side code. It's a tool to be used in dev (node.js).

File: README.md [hunk 6]
```diff
@@ -1,2 +1,61 @@
...
+Multiple configurations can be combined together with function composition. For example:
+
+```js
+// next.config.js
+const withCSS = require('@zeit/next-css');
+const withRuntimeEnv = require('@moxy/next-runtime-env');
+
+module.exports = withCSS(
+    withRuntimeEnv()({
+        cssModules: true, // this options will be passed to withCSS plugin through nextConfig
```
Comment (satazor) [inline README.md:51]: The first time I read this, I was assuming that `cssModule` was an option of this plugin. Perhaps, in this example, should we give an example where we pass options of our own plugin?

File: README.md [hunk 7]
```diff
@@ -1,2 +1,61 @@
...
+    }),
+);
+```
+
+### API
+
+| Option         | Description                                                        | Type    | Default   |
+|----------------|--------------------------------------------------------------------|---------|-----------|
```
Comment (satazor) [inline README.md]: Nifty pick: you don't need to define `---------------------------` with exact length. It's preferable to use it normally to avoid having a lot of lines in a git diff when a new option is added, especially when it has a larger length in either the name or description. 

e.g.: https://raw.githubusercontent.com/moxystudio/babel-preset-moxy/master/README.md and check the table there

File: index.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * NextJS plugin to pass environment variables to NextJS's runtime configuration (`publicRuntimeConfig` and `serverRuntimeConfig`)
+ *
+ * @param {Object} options
+ * @param {String} options.publicPrefix     Prefix of variables to lookup and then pass to publicRuntimeConfig
+ * @param {String} options.serverPrefix     Prefix of variables to lookup and then pass to serverRuntimeConfig
+ * @param {String} options.removePrefixes   Option to remove prefix when passing variables to runtime config
+ */
+
+const runtimeEnv = (options = {}) => (nextConfig = {}) => {
```
Comment (satazor) [inline index.js:29]: Do we need the default value for `nextConfig`? I think next always passes us a plain-object.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    "dotenv",
+    "runtime"
+  ],
+  "license": "MIT",
+  "repository": {
+    "type": "git",
+    "url": "git@github.com:moxystudio/next-runtime-env.git"
+  },
+  "scripts": {
+    "clean": "rm -rf node_modules coverage",
```
Comment (satazor) [inline package.json]: I don't think we need the clean command.

Moreover, we are missing the commitlint, standard-version, and perhaps other tooling? Check out other MOXY's packages: https://github.com/moxystudio/babel-preset-moxy/blob/master/package.json#L24

**Review by satazor — DISMISSED** [review]
Overall looks good.

In regards to the `.gitignore`, we should only have the files specific to this package and not OS specific files there. We can discuss more in detail personally.

Moreover, there are other personal opinions related to using separates files for husky, lintstaged, etc that we should discuss as well.

Lets talk tomorrow?

File: package.json (hunk 1 above)
Comment (ivolimasilva) [reply package.json:3]: Sure, added command for `release`.

File: README.md (hunk 2 above)
Comment (ivolimasilva) [reply README.md]: Rephrased! Please check again, phrasing isn't my strongest skill 😢

File: README.md (hunk 3 above)
Comment (ivolimasilva) [reply README.md]: I like it! Changed it!

File: README.md (hunk 5 above)
Comment (ivolimasilva) [reply README.md]: Removed then.

File: README.md (hunk 7 above)
Comment (ivolimasilva) [reply README.md]: Yeah, I was using an online tool to build tables and by default it defined the number of `-` with exact length. I also prefer without it since when changing in the future, it would probably not be correct.

Removed the unnecessary `-`.

File: package.json (hunk 9 above)
Comment (ivolimasilva) [reply package.json]: Removed and added the missing tooling. Will leave this conversation open so you don't forget to check if it's correct now. Thanks.

File: README.md (hunk 6 above)
Comment (ivolimasilva) [reply README.md:51]: Passed some options and also change the above example since it was incorrect.

Using

```js
module.exports = withRuntimeEnv();
```

was incorrect because NextJS calls this file with `(phase, nextConfig)` and not with `(nextConfig)`. Because of this, when using only this plugin, we have to use like this:

```js
module.exports = withRuntimeEnv()();
```

However this means we will be using an empty `nextConfig`. This is not problematic since we're not using anything from the default configuration. However, a more correct way of using is:

```js
module.exports = (phase, nextConfig) => withRuntimeEnv({ ...options })(nextConfig);
```

But that is strange for a single plugin.

File: index.js (hunk 8 above)
Comment (ivolimasilva) [reply index.js:29]: It depends (see discussion above).

If we're using like this:

```js
module.exports = withRuntimeEnv()();
```

Then the `nextConfig` is `undefined`. I believe it's best to have this default value because of this example (same thing happens in all of **zeit**'s own plugins).
